### DISCUSSION TITLE
Audit.NET Doesn't work with entity framework core 5

### DISCUSSION THREAD

**Discussion by ApexunitedEgypt** [thepirat000/Audit.NET#370] [Q&A]
**Describe the bug**
Adudit.NET was working fine and then When I upgraded the project it didn't work and the error message was:
Could not load type Microsoft.EntityFrameworkCore.RawSqlString from assembly Microsoft.EntityFrameworkCore.Relational

**To Reproduce**
Steps or code to reproduce the behavior.

**Expected behavior**

**Libraries (specify the Audit.NET extensions being used including version):**
For example:
 - Audit.NET: 16.4.4
 - Audit.NET.SqlServer: 14.6.0

**Target .NET framework:**
For example:
 - .NET Standard 2.1
 - .NET Framework 5.0.2

**Additional context**

**Comment by thepirat000** [comment]
You must use the same version number for Audit.NET and Audit.NET.SqlServer
Is that a typo?

**Comment by tessierp** [comment]
Same problem here, after upgrading to .NET CORE 5.0, what I have implemented before using Audit.NET no longer works.

I'm making use of DbContextHelper and IAuditDbContext. I use the default PostgreSQL provider, not AuditNet's one, this worked fin in the past but ever since I upgrade to .NET CORE 5.0, it no longer works.

```c# 
public class DbContextAuditNet : DbContextBase
    {
        private static DbContextHelper _helper = new DbContextHelper();
        private readonly IAuditDbContext _auditContext;

        #region Initializers

        public DbContextAuditNet() { }

        public DbContextAuditNet(DbContextOptions options)
            : base(options)
        {
            _auditContext = new DefaultAuditContext(this);
            _helper.SetConfig(_auditContext);
        }

        public DbContextAuditNet(
            DbContextOptions options,
            bool enableAuditTrails,
            string defaultSchema = "")
            : this(options)
        {
            EnableAuditTrails = enableAuditTrails;
            DefaultSchema = defaultSchema;
        }

        #endregion

        #region Properties

        public bool EnableAuditTrails { get; private set; } = false;

        #endregion

        #region Overrides

        public override async Task<int> SaveChangesAsync(CancellationToken cancellationToken = default(CancellationToken))
        {
            if (EnableAuditTrails)
            {
                ChangeTracker.DetectChanges();
                return await _helper.SaveChangesAsync(_auditContext, () => base.SaveChangesAsync(cancellationToken));
            }

            return await base.SaveChangesAsync(cancellationToken);
        }

        public override int SaveChanges()
        {
            if (EnableAuditTrails)
            {
                ChangeTracker.DetectChanges();
                return _helper.SaveChanges(_auditContext, () => base.SaveChanges());
            }

            return base.SaveChanges();
        }

        #endregion

        #region Protected Methods

        protected override void OnModelCreating(ModelBuilder modelBuilder)
        {
            base.OnModelCreating(modelBuilder);

            if (EnableAuditTrails)
            {
                modelBuilder.Entity<AuditLog>(e =>
                {
                    e.HasKey(a => a.AuditId);
                });
            }
        }

        #endregion
    }
}
```

The class above receives its context from this fonction : 

```c#
        protected override SomeDbContext CreateContext(DbSettings databaseSettings)
        {
            var dbContextOptionBuilder = new DbContextOptionsBuilder<SomeDbContext>();

            if (databaseSettings.Provider == IO.Database.Enums.DbProvider.PostgreSQL)
            {
                dbContextOptionBuilder.UseNpgsql(databaseSettings.ConnectionString);
            }
            else
            {
                dbContextOptionBuilder.UseSqlServer(databaseSettings.ConnectionString);
            }

            return new SomeDbContext(dbContextOptionBuilder.Options, databaseSettings.EnableAudits, databaseSettings.Schema);
        }
```

SomeDbContext derives from DbContextAuditNet and DbContextAuditNet derives from Microsoft's DbContext.

All this worked before 5.0.

**Comment by thepirat000** [comment]
Please share your csproj, at least the versions of all the packages you're referencing

**Comment by tessierp** [comment]
Of course : 

    <PackageReference Include="Microsoft.EntityFrameworkCore" Version="5.0.3" />
    <PackageReference Include="Microsoft.EntityFrameworkCore.Design" Version="5.0.3">
        <PrivateAssets>all</PrivateAssets>
        <IncludeAssets>runtime; build; native; contentfiles; analyzers; buildtransitive</IncludeAssets>
    </PackageReference>
    <PackageReference Include="Microsoft.EntityFrameworkCore.Relational" Version="5.0.3" />
    <PackageReference Include="Microsoft.EntityFrameworkCore.SqlServer" Version="5.0.3" />
    <PackageReference Include="Microsoft.Extensions.Configuration.FileExtensions" Version="5.0.0" />
    <PackageReference Include="Microsoft.Extensions.Configuration.Json" Version="5.0.0" />
    <PackageReference Include="System.Data.SqlClient" Version="4.8.2" />
    <PackageReference Include="System.Text.Json" Version="5.0.1" />

    <PackageReference Include="Audit.EntityFramework.Core" Version="16.4.4" />


    <PackageReference Include="Autofac" Version="6.1.0" />
    <PackageReference Include="Microsoft.EntityFrameworkCore.SqlServer" Version="5.0.3" />
    <PackageReference Include="Microsoft.Extensions.Configuration" Version="5.0.0" />
    <PackageReference Include="Microsoft.Extensions.Configuration.Abstractions" Version="5.0.0" />
    <PackageReference Include="Microsoft.Extensions.Configuration.Binder" Version="5.0.0" />
    <PackageReference Include="Microsoft.Extensions.Configuration.FileExtensions" Version="5.0.0" />
    <PackageReference Include="Microsoft.Extensions.Configuration.Json" Version="5.0.0" />
    <PackageReference Include="Npgsql.EntityFrameworkCore.PostgreSQL" Version="5.0.2" />

This is a list of the packages for the libraries dealing with the database.

**Comment by thepirat000** [comment]
I still cannot reproduce. Could you provide a minimal runnable project that fails?

I've tested with the following csproj and had no problems to audit an EF dbcontext:

```xml
<Project Sdk="Microsoft.NET.Sdk.Web">
  <PropertyGroup>
    <TargetFramework>net5.0</TargetFramework>
  </PropertyGroup>

  <ItemGroup>
  <PackageReference Include="Microsoft.EntityFrameworkCore" Version="5.0.3" />
  <PackageReference Include="Microsoft.EntityFrameworkCore.Design" Version="5.0.3">
    <PrivateAssets>all</PrivateAssets>
    <IncludeAssets>runtime; build; native; contentfiles; analyzers; buildtransitive</IncludeAssets>
  </PackageReference>
  <PackageReference Include="Microsoft.EntityFrameworkCore.Relational" Version="5.0.3" />
  <PackageReference Include="Microsoft.EntityFrameworkCore.SqlServer" Version="5.0.3" />
  <PackageReference Include="Audit.EntityFramework.Core" Version="16.4.4" />
  </ItemGroup>
</Project>
```

Do you get any warning when restoring packages or building?

**Comment by tessierp** [comment]
No, no warning or errors at all. I have debugged live and get into the SaveChangesAsync and audit is enabled and these two lines are called

                ChangeTracker.DetectChanges();
                return await _helper.SaveChangesAsync(_auditContext, () => base.SaveChangesAsync(cancellationToken));

Yet nothing is written into the AuditLog table. Is it possible it could be a Schema issue? I am writing this in a different schema.

**Comment by tessierp** [comment]
OK so the issue is not caused by using something different from the default schema.

**Comment by thepirat000** [comment]
@tessierp loooks like your issue is not related with the original problem with @ApexunitedEgypt 
Do you get the 'Could not load type' exception?

Provide a complete example, otherwise I cannot help. How do you configured the data provider for audit log saving?

**Comment by tessierp** [comment]
I didn't get any exception otherwise I would have caught it. I am able to write to the database, the SaveChangesAsync runs without any issue. As you can see in the code these two lines are called : 

            _auditContext = new DefaultAuditContext(this);
            _helper.SetConfig(_auditContext);

where "this" is my context, does not derive from AuditDbContext. The reason is I have a generic one I need to use hence why I am using DefaultAuditContext and passing it my context object and using the helper to set the context..

The context is created like this : 
`
        protected override MyDbContext CreateContext(CpiirDbSettings databaseSettings)
        {
            var dbContextOptionBuilder = new DbContextOptionsBuilder<MyDbContext>();

            if (databaseSettings.Provider == IO.Database.Enums.DbProvider.PostgreSQL)
            {
                dbContextOptionBuilder.UseNpgsql(databaseSettings.ConnectionString);
            }
            else
            {
                dbContextOptionBuilder.UseSqlServer(databaseSettings.ConnectionString);
            }

            return new MyDbContext(dbContextOptionBuilder.Options, databaseSettings.EnableAudits, databaseSettings.Schema);
        }
`

The PostgreSQL line gets called.

I can't provide a complete example as it would mean too much code at the moment. All I can say is the only changes made was upgrading to NET CORE 5.0. Is it still possible to pass my context to DefaultAuditContext and then call DbContextHelper.SetConfig and passing it the AuditContext created? That is the method I am using which worked before.

Notice that I do not use the Audit.NET PostgreSQL provider, but the generic one. I was using MSSQL before and used the generic one provided by Microsoft without any issues.

Perhaps if I knew what you are doing I could see the difference between what I do and what you do it can help me see what is different.

**Comment by thepirat000** [comment]
You can always clone the repo and import the project to reference it locally on your solution. 

But still you are not showing how do you configure the output of the audit. Normally you should have a `Audit.Core.Configuration.Setup().UseEntityFramework()` call or `Audit.Core.Configuration.DataProvider = new ...` call. Can you share that?

**Comment by tessierp** [comment]
Perhaps I am wrong but I thought that is what DbContextHelper was doing when I called SetConfig and passing it the audit context. 

I did try this
Audit.Core.Configuration.Setup().UseEntityFramework(config => config.UseDbContext<MyDbContext>(dbContextOptionBuilder));
but still can't get it to work.

**Comment by thepirat000** [comment]
No, the output is independently configured. Check the [data providers](https://github.com/thepirat000/Audit.NET#data-providers) section on the readme.

By default (if you don't assign any dataprovider) it will log to .json files on the current folder.

If you want to store the audit logs using the same dbcontext you are auditing, you need to provide the configuration depending on how do you want to save the audit logs. Basically you need to provide a way to map the audited entities to the audit log entities.

Some examples [here](https://github.com/thepirat000/Audit.NET/blob/master/src/Audit.EntityFramework/README.md#ef-provider-configuration-examples)

**Comment by tessierp** [comment]
I will check it out, thank you!

I'm presuming this Configuration step can only be done at the application's Startup phase?

**Comment by thepirat000** [comment]
That's correct, it should be done only once

**Comment by ApexunitedEgypt** [comment]
Thank you My problem has been solved.. 
I installed the same version of Audit.NET and Audit.NET.SqlServer (16.4.4).

Thank you.
